src/audio/pipeline.c

```diff
@@ -285,26 +285,31 @@ static void pipeline_set_params(struct comp_dev *comp,
 				struct sof_ipc_pcm_params *params,
 				int dir)
 {
-	struct list_item *buffer_list;
-	struct list_item *clist;
+	struct list_item *buffer_list = comp_buffer_list(comp, dir);
+	struct list_item *clist = buffer_list->next;
+	struct list_item *curr;
 	struct comp_buffer *buffer;
 	int i;
 
 	/* set comp params */
 	comp->direction = params->params.direction;
 
-	/* set buffer params */
-	buffer_list = comp_buffer_list(comp, dir);
+	while (clist != buffer_list) {
```

Comment:
`for (clist = buffer_list->next; clist != buffer_list; clist = clist->next)` eliminates the need for `curr`

Reply:
@lyakh Then you will access buffer after invalidating it. With my approach I can limit number of buffer locks in params and prepare calls of audio components to only the ones, that in fact change the format.

Reply:
@tlauda are you talking about invalidation, added in a separate patch? Which one is this?

Reply:
@lyakh It's done in buffer_lock() function.

Reply:
@tlauda then sorry, I don't understand how replacing "while" with "for" would change anything about locking / cache invalidation. The only difference I see is that in your case the `.next` pointer is dereferenced before invalidating caches and in the existing / my case - after. But you anyway read in `buffer` into the cache in `buffer_set_params()` when checking `buffer->hw_params_configured`

Reply:
@lyakh And after that I'm doing buffer_unlock(), which will do writeback and invalidate. Now from this point I don't want to load the same buffer again into the local cache, but use the fact that the memory is invalidated and automatically load new data (without doing explicit invalidation) in the places, where I'm just reading parameters without doing any modifications to the buffer. This kind of solution works OK with current FW flows, because buffer params are only modified during params and prepare calls.

Reply:
@tlauda Right, as long as you do invalidate caches in unlock, which I don't understand why it's necessary. And do you expect all (or at least a significant number of) buffers be in cache when you enter that function? If they aren't in cache, then the same thing happens anyway: a buffer is read into cache to get its `.next` pointer, then it's invalidated, then it's read in again and invalidated again.

Reply:
@lyakh As I stated in my previous comment - it's necessary in order to avoid calling buffer_lock/unlock in component_params/prepare calls, where buffer's params are only read, not modified. I think I see your point here and getting next pointer should be moved after buffer_lock.